### PR TITLE
Use dynamic code loading in ssl_compat and time_compat

### PR DESCRIPTION
**Pull request by dcorbacho**
Based on dynamic code load in startup/runtime. See https://github.com/rabbitmq/rabbitmq-server/issues/346#issuecomment-169980207

Solves https://github.com/rabbitmq/rabbitmq-server/issues/346 and https://github.com/rabbitmq/rabbitmq-server/issues/347

### CODE REVIEW

**Comment by dumbbell** [comment]
For me (Erlang 18.2.2), RabbitMQ doesn't start: `time_compat:monotonic_time/0` enters an infinite loop trying to update/reload the module. I'm going to debug this and give more informations.

File: src/time_compat.erl [hunk 1]
```diff
@@ -43,160 +43,242 @@
...
+-version_support(
+   [{monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 0},
+    {monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 1},
+    {erlang_system_time, erlang_system_time_pre_18, erlang_system_time_post_18, 0},
+    {erlang_system_time, erlang_system_time_pre_18, erlang_system_time_post_18, 1},
+    {os_system_time, os_system_time_pre_18, os_system_time_post_18, 0},
+    {os_system_time, os_system_time_pre_18, os_system_time_post_18, 1},
+    {time_offset, time_offset_pre_18, time_offset_post_18, 0},
+    {time_offset, time_offset_pre_18, time_offset_post_18, 1},
+    {convert_time_unit, convert_time_unit_pre_18, convert_time_unit_post_18, 0},
```
Comment (dumbbell) [inline src/time_compat.erl]: The infinite loop comes from this line: `convert_time_unit` arity is 3, not 0. Thus, convert_time_unit is not replaced.

File: src/time_compat.erl [hunk 2]
```diff
@@ -43,160 +43,242 @@
...
+monotonic_time_post_18(Unit) ->
     try
-	erlang:monotonic_time()
+        erlang:monotonic_time(Unit)
     catch
-	error:undef ->
-	    %% Use Erlang system time as monotonic time
-	    erlang_system_time_fallback()
+        error:badarg ->
+            erlang:error(badarg, [Unit])
```
Comment (dumbbell) [inline src/time_compat.erl]: Why do you catch exceptions from the official API? I believe the compat modules should not change anything, it should be a transparent wrapper of the API.

File: src/code_version.erl [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                false ->
+                    Tree
+            end
+    end.
+
+rename_abstract_functions(ToRename, ToName) ->
+    fun(Tree, {Name, _Arity} = Function) ->
+            case lists:member(Function, ToRename) of
+                true ->
+                    FunctionName = proplists:get_value(Name, ToName),
```
Comment (dumbbell) [inline src/code_version.erl]: You ignore the arity of the function about to be renamed. `ToName`, returned by `get_name_pairs/2`, looks like:
```erlang
ToName=[{monotonic_time_post_18,monotonic_time}, % Comes from {monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 0}
        {monotonic_time_post_18,monotonic_time}, % Comes from {monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 1}
        % ...
```

It doesn't contain the arity information either, but this is important: here you pick the new name from the first matching entry, ie. always the `{monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 0}` attribute, even if the new name is different for the `/1` function.

File: src/time_compat.erl (hunk 2 above)
Comment (dcorbacho) [inline src/time_compat.erl]: All those try/catch clauses were in the previous implementation, so I left them there. I can't think of why they could be needed, so if we both agree they can be removed.

Comment (dcorbacho) [inline src/time_compat.erl]: Okay, I found a case where those try/catch do 'something'. I think in the case of `convert_time_unit_fallback` the author wanted to return the exact badarg error the system call would return, that is, with the original function call arguments. In others, like `monotonic_time`, it doesn't change anything. What shall we do @dumbbell ?

**Comment by michaelklishin** [comment]
Merging per discussion in #346 and another round of QA on R16B03 and 18.2.1, with and without HiPE.

**Pull request merged**